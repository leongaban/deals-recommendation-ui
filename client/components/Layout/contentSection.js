import React from 'react';

// MUI Components
import Paper from '@material-ui/core/Paper';
import { withStyles } from '@material-ui/core/styles';

// Styles
import { styles as paperStyles } from 'components/Styles/paper';

const ContentSection = ({ children, classes }) => (
  <Paper className={classes.paperDefault}>{children}</Paper>
);

export default withStyles(paperStyles)(ContentSection);
