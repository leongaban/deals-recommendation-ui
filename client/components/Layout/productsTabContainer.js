import React from 'react';

// Material-UI
import Typography from '@material-ui/core/Typography';

// Components
import Table from 'components/Layout/table';

// Constants
import { TABS_DEFINITIONS } from 'constants/tabs';

function TabContainer(props) {
  return (
    <Typography component="div">
      {props.children}
    </Typography>
  );
}

const ProductsTabContainer = (props) => {
  const { value } = props;
  const tabDef = TABS_DEFINITIONS[value];

  // Abort if we can't find the value.
  if (!tabDef) return null;

  // Get the list for this products/tab type.
  const list = props[tabDef.list];

  // Determine if this is the "all" list or not.
  const allList = value === 0;

  return (
    <TabContainer>
      <Table {...allList} list={list} />
    </TabContainer>
  );
};

export default (ProductsTabContainer);
