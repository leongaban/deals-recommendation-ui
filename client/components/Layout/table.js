/* eslint-disable no-param-reassign */
/* eslint-disable no-return-assign */
import React, { Component } from 'react';
import { connect } from 'react-redux';

// Actions
import { updateRules as putRules } from 'actions/Rules';
import { setModal as setModalAction } from 'actions/Modals';
import { toggleNotification as displayToggleNotification } from 'actions/Notifications';

// Components
import Thead from 'components/Layout/Partials/thead';
import TableRowsContainer from 'components/Layout/tableRowsContainer';
import TableRow from 'components/Layout/tableRow';

// Utils
import { cleanMapStateToProps } from 'utils/redux';
import {
  changeRuleType,
  changeEnabledStatus,
  changedRanks,
  checkLimits,
  findListType
} from 'utils/rules';

// Strings
import { DELETE_PRODUCT_MODAL } from 'constants/modals';

// Material-UI
import { withStyles } from '@material-ui/core/styles';

// Styles
import { styles } from './table.styles';

class Table extends Component {
  static getDerivedStateFromProps(nextProps) {
    const { list } = nextProps;
    const listType = findListType(list);
    return { list, listType };
  }

  constructor(props) {
    super(props);

    // List comes from props and is set to state.
    const { list } = props;

    this.state = {
      list,
      listType: ''
    };

    this.handleChangeType = this.handleChangeType.bind(this);
    this.handleToggleEnable = this.handleToggleEnable.bind(this);
    this.handleSortUpdate = this.handleSortUpdate.bind(this);
    this.handleDelete = this.handleDelete.bind(this);
  }

  handleChangeType(product, event) {
    const {
      rulesSort,
      update,
      list,
      rulesDefault: dList,
      rulesOverride: oList
    } = this.props;

    const { listType } = this.state;
    const { ruleID, rank } = product;
    const { value: type } = event.target;

    const rulePackage = {
      listType, list, type, rank, rulesSort, ruleID, dList, oList
    };

    const newRules = changeRuleType(rulePackage);
    update(newRules);
  }

  handleToggleEnable(event) {
    const {
      rulesDefault,
      rulesOverride,
      rulesExcluded,
      rulesSort,
      update,
      list,
      toggleNotification
    } = this.props;

    const { value: ruleID } = event.target;
    const rule = list.filter(i => ruleID === i.ruleID).pop();
    const type = rule.checked.charAt(0);
    const canToggle = checkLimits(type, rulesDefault, rulesOverride, rulesExcluded);
    const { enabled } = rule;

    // If the list has room OR if we are disabling then update:
    if (canToggle || enabled) {
      const newRules = changeEnabledStatus(list, rulesSort, ruleID);
      update(newRules);
    } else {
      const errMsg = `You have hit the enabled ${rule.checked.toLowerCase()} limit.`;
      toggleNotification(errMsg, 'error');
    }
  }

  /**
   * Updates the sort order for an entity group.
   * @param {{ oldIndex: number, newIndex: number }}
   */
  handleSortUpdate({ oldIndex, newIndex }) {
    const { update, rulesSort } = this.props;
    const { listType } = this.state;

    const list = [
      ...this.state.list
    ];

    const [item] = list.splice(oldIndex, 1);
    list.splice(newIndex, 0, item);

    const rerankedList = list.map((rule, index) => {
      rule.rank = index + 1;
      return rule;
    });

    const newRules = changedRanks(listType, rerankedList, rulesSort);

    update(newRules);

    this.setState({
      ...this.state, list
    });
  }

  handleDelete({ description, ruleID }) {
    const { setModal } = this.props;
    setModal(DELETE_PRODUCT_MODAL, { description, ruleID });
  }

  renderList() {
    const { classes, setModal } = this.props;
    const { list } = this.state;

    return list.map((rule, index) => (
      <TableRow
        key={rule.ruleID}
        index={index}
        classes={classes}
        changeType={this.handleChangeType}
        changeEnable={this.handleToggleEnable}
        deleteRule={this.handleDelete}
        setModal={setModal}
        style={styles}
        rule={rule}
      />
    ));
  }

  render() {
    return (
      <table>
        <Thead />
        <TableRowsContainer onSortEnd={this.handleSortUpdate}>
          { this.renderList() }
        </TableRowsContainer>
      </table>
    );
  }
}

export const mapDispatchToProps = dispatch => ({
  update: (...args) => dispatch(putRules(...args)),
  setModal: (...args) => dispatch(setModalAction(...args)),
  toggleNotification: (...args) => dispatch(displayToggleNotification(...args))
});

const mapStateToProps = cleanMapStateToProps([
  'rulesDefault',
  'rulesOverride',
  'rulesExcluded',
  'rulesSort'
]);

export const TableJest = Table;

export default connect(mapStateToProps, mapDispatchToProps)(withStyles(styles)(Table));
