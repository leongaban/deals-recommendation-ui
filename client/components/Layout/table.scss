.table-container {
  margin-top: 20px;
  width: 100%;

  table { width: 100%; color: #000; }

  th {
    padding: 10px;
    font-weight: 700;
    border-bottom: 1px solid $darkGray;

    &:first-child,
    &:nth-child(2) {
      text-align: left;
    }

    &:nth-child(3) {
      text-align: center;
    }
  }

  td {
    margin: 5px 0;
    padding: 10px;
    border-bottom: 1px solid $darkGray;
  }
}

.description {
  min-width: 200px;
  max-width: 200px;
  text-align: left;

  section {
    span {
      float: left;
      width: 100%;
    }
  }
}

.dropdown-arrow,
.dropup-arrow { width: 22px; height: 15px; }

.dropdown-arrow,
.dropup-arrow,
.pencil-ico,
.trash-ico {
  @include transition(opacity, .3s, ease-in-out);
  opacity: 0.6;
  cursor: pointer;
  &:hover { opacity: 1; }
}

.dropdown-arrow { background-image: url('static/imgs/down.png'); }

.dropup-arrow { background-image: url('static/imgs/up.png'); }

.pencil-ico,
.trash-ico {
  background-position: 50% 50%;
  background-repeat: no-repeat;
}

.pencil-ico {
  width: 18px;
  height: 18px;
  background-image: url('static/imgs/pencil.png');
}

.trash-ico {
  width: 14px;
  height: 18px;
  background-image: url('static/imgs/trash.png');
}
