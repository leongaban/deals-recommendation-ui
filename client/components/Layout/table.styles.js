export const styles = theme => ({
  tr: {
    width: '100%'
  },
  td: {
    margin: '5px 0',
    padding: '5px 0',
    height: '20px',
    backgroundColor: '#fff',
    borderTop: '1px solid #4a4a4a',
    borderBottom: '1px solid #4a4a4a'
  },
  formControl: {
    display: 'flex',
    float: 'left',
    margin: theme.spacing.unit * 3
  },
  group: {
    float: 'left',
    flexDirection: 'row'
  }
});
