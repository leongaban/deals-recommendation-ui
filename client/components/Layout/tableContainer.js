/* eslint-disable object-curly-newline */
import React, { Component } from 'react';
import { connect } from 'react-redux';

// Actions
import { getRules as fetchRules } from 'actions/Rules';

// Material-UI
import { withStyles } from '@material-ui/core/styles';
import AppBar from '@material-ui/core/AppBar';
import Tabs from '@material-ui/core/Tabs';
import Tab from '@material-ui/core/Tab';

// Utils
import { cleanMapStateToProps } from 'utils/redux';
import { isArrayEmpty } from 'utils/common';

// Components
import ProductsTabContainer from 'components/Layout/productsTabContainer';

// Constants
import { TABS_DEFINITIONS } from 'constants/tabs';

// Styles
import { styles } from './tableContainer.styles';

class TableContainer extends Component {
  constructor(props) {
    super(props);

    const {
      getRules,
      rulesAll
    } = props;

    this.state = { value: 0 };

    if (isArrayEmpty(rulesAll)) {
      getRules();
    }

    this.handleChangeTab = this.handleChangeTab.bind(this);
  }

  handleChangeTab(event, value) {
    this.setState({ value });
  }

  render() {
    const { classes } = this.props;
    const { value } = this.state;

    const { tabRoot, tabsRoot, tabSelected, tabsIndicator } = classes;
    const tabsClasses = { root: tabsRoot, indicator: tabsIndicator };
    const tabClasses = { root: tabRoot, selected: tabSelected };

    return (
      <div className={classes.root}>
        <AppBar position="static">
          <Tabs value={value} classes={tabsClasses} onChange={this.handleChangeTab}>
            {TABS_DEFINITIONS.map(({ label }) => (
              <Tab key={label} classes={tabClasses} label={label} />
            ))}
          </Tabs>
        </AppBar>
        <ProductsTabContainer value={value} {...this.props} />
      </div>
    );
  }
}

export const mapDispatchToProps = dispatch => ({
  getRules: (...args) => dispatch(fetchRules(...args))
});

const mapStateToProps = cleanMapStateToProps([
  'rulesAll',
  'rulesOverride',
  'rulesDefault',
  'rulesExcluded'
]);

export const TableContainerJest = TableContainer;

export default connect(mapStateToProps, mapDispatchToProps)(withStyles(styles)(TableContainer));
