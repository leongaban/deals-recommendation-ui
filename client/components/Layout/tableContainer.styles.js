export const styles = theme => ({
  root: {
    flexGrow: 1
  },
  tabsRoot: {
    border: '1px solid #4a4a4a',
    backgroundColor: '#f0f0f0',
    boxShadow: 'none'
  },
  tabsIndicator: {
    color: '#f44336',
    backgroundColor: '#f44336'
  },
  tabRoot: {
    textTransform: 'initial',
    minWidth: 72,
    fontWeight: theme.typography.fontWeightRegular,
    marginRight: theme.spacing.unit * 4,
    color: '#f44336',
    fontFamily: [
      'Roboto',
      'Arial',
      'sans-serif'
    ].join(','),
    '&:hover': {
      color: '#c9372c',
      opacity: 1
    },
    '&$tabSelected': {
      color: '#c9372c'
    },
    '&:focus': {
      color: '#c9372c'
    }
  },
  tabSelected: {},
  typography: {
    padding: theme.spacing.unit * 3
  }
});
