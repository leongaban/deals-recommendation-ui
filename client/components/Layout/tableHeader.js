import React, { Component } from 'react';
import { connect } from 'react-redux';

// Actions
import { updateSortRule as putSortRule } from 'actions/Rules';

// Utils
import { cleanMapStateToProps } from 'utils/redux';

// Constants
import {
  LIFT,
  WEIGHTEDLIFT,
  WEIGHTED_LIFT
} from 'copy/Components/tableHeader';

// Material-UI
import { withStyles } from '@material-ui/core/styles';
import Radio from '@material-ui/core/Radio';
import RadioGroup from '@material-ui/core/RadioGroup';
import FormControlLabel from '@material-ui/core/FormControlLabel';

// Styles
import { styles } from './tableHeader.styles';

class TableHeader extends Component {
  static getDerivedStateFromProps(nextProps) {
    const { rulesSort } = nextProps;
    return { rulesSort };
  }

  constructor(props) {
    super(props);

    this.state = {
      rulesSort: 'Lift'
    };

    this.handleSortChange = this.handleSortChange.bind(this);
  }

  handleSortChange(event) {
    const { updateSortRule } = this.props;
    const { value: sortRule } = event.target;
    updateSortRule({ sortRule });
  }

  render() {
    const {
      classes, rulesDefault, rulesOverride, rulesExcluded
    } = this.props;
    const { rulesSort } = this.state;

    const defaultEnabled = rulesDefault.filter(i => i.enabled === true);
    const overrideEnabled = rulesOverride.filter(i => i.enabled === true);
    const excludedLen = rulesExcluded.length;

    return (
      <header className="table-header">
        <h2 className={classes.blackColor}>Sort Rules</h2>

        <div className="sort-options">
          <RadioGroup
            aria-label="lift"
            name="lift"
            className={classes.group}
            value={rulesSort}
            onChange={e => this.handleSortChange(e)}
          >
            <FormControlLabel value={LIFT} control={<Radio />} label={LIFT} />
            <FormControlLabel value={WEIGHTEDLIFT} control={<Radio />} label={WEIGHTED_LIFT} />
          </RadioGroup>
        </div>

        <section className="limits">
          <ul>
            <li>Enabled limits:</li>
            <li>Overrides ({overrideEnabled.length}/4)</li>
            <li>Defaults ({defaultEnabled.length}/8)</li>
            <li>Excluded ({excludedLen}/8)</li>
          </ul>
        </section>
      </header>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  updateSortRule: (...args) => dispatch(putSortRule(...args))
});

const mapStateToProps = cleanMapStateToProps([
  'rulesDefault',
  'rulesOverride',
  'rulesExcluded',
  'rulesSort'
]);

export const TableHeaderJest = TableHeader;

export default connect(mapStateToProps, mapDispatchToProps)(withStyles(styles)(TableHeader));
