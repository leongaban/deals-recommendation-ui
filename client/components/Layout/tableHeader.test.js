// Test utils
import { testCommonComponentAttrs } from 'utils/tests';

// Constants
import { LIFT } from 'copy/Components/tableHeader';

// Components
import { TableHeaderJest } from 'components/Layout/tableHeader';

const mockUpdateSortRule = jest.fn(() => Promise.resolve());

jest.mock('actions/Rules', () => ({
  updateSortRule: () => mockUpdateSortRule()
}));

const mockProps = {
  classes: {
    root: 'root'
  },
  getRules: mockUpdateSortRule,
  rulesDefault: [],
  rulesOverride: [],
  rulesExcluded: [],
  rulesSort: LIFT
};

describe('<TableHeaderJest /> component', () => {
  testCommonComponentAttrs(TableHeaderJest, mockProps);
});
