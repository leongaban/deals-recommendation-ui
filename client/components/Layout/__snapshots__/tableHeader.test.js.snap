// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TableHeaderJest /> component when rendering should render a component matching the snapshot 1`] = `
<header
  className="table-header"
>
  <h2>
    Sort Rules
  </h2>
  <div
    className="sort-options"
  >
    <RadioGroup
      aria-label="lift"
      name="lift"
      onChange={[Function]}
      value="Lift"
    >
      <WithStyles(FormControlLabel)
        control={<WithStyles(Radio) />}
        label="Lift"
        value="Lift"
      />
      <WithStyles(FormControlLabel)
        control={<WithStyles(Radio) />}
        label="Weighted Lift"
        value="WeightedLift"
      />
    </RadioGroup>
  </div>
  <section
    className="limits"
  >
    <ul>
      <li>
        Enabled limits:
      </li>
      <li>
        Overrides (
        0
        /4)
      </li>
      <li>
        Defaults (
        0
        /8)
      </li>
      <li>
        Excluded (
        0
        /8)
      </li>
    </ul>
  </section>
</header>
`;
